File: C#/0589. N-ary Tree Preorder Traversal.cs
/*
// Definition for a Node.
public class Node {
    public int val;
    public IList<Node> children;

    public Node(){}
    public Node(int _val,IList<Node> _children) {
        val = _val;
        children = _children;
}
*/
public class Solution {
    public IList<int> Preorder(Node root) {
        IList<int> rep=new List<int>();
        if (root==null){
            return rep;
        }
        Stack<Node> nodeStack=new Stack<Node>();
        nodeStack.Push(root);
        while(nodeStack.Count!=0){
            Node curNode=nodeStack.Pop();
            rep.Add(curNode.val);
            for (int i=curNode.children.Count-1;i>=0;i--){
                nodeStack.Push(curNode.children[i]);
            }
        }
        return rep;
    }
}

File: 20190519python/0589. N-ary Tree Preorder Traversal.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""
class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        if not root:
            return []
        rep=[]
        root_list=[root]
        while root_list:
            cur_root=root_list.pop()
            rep.append(cur_root.val)
            for child in cur_root.children[::-1]:
                root_list.append(child)
        return rep
        
        
#================================================================
"""
# Definition for a Node.
class Node:
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""
class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        if not root:
            return []
        rep=[]
        self.sub_preorder(root,rep)
        return rep
        
    def sub_preorder(self,root,rep):
        if not root:
            return
        rep.append(root.val)
        for child in root.children:
            self.sub_preorder(child,rep)
        return
